bootstrap/src/lib/cache/rateLimit.ts

```diff
@@ -0,0 +1,140 @@
+import { logger } from '@chainlink/external-adapter'
+import { DEFAULT_CACHE_MAX_AGE } from './redis'
+
+const MAX_AGE_ALLOWED = 1000 * 60 * 2
+
+type RateLimitOptions = {
+  groupMaxAge: number
+  groupId: string
+  id: string
+  totalCapacity: number
+}
+
+export interface RateLimit {
+  isEnabled: () => boolean
+  getParticipantMaxAge: (participantId: string) => number | false
+  incrementParticipantHeartbeat: (participantId: string, cost?: number) => number | false
+}
+
+type Participant = {
+  cost: number
+  heartbeat: number
+  lastSeen: number
+}
+
+const totalHeartbeat = new Map()
+let participants: Map<string, Participant> = new Map()
+
+export const makeRateLimit = (options: RateLimitOptions): RateLimit => {
+  const minHeartbeat = 0
+  const safeCapacity = options.totalCapacity * 0.9
+
+  const _isEnabled = (): boolean => {
+    return !!options.groupId && !!options.totalCapacity
+  }
+
+  // Weight normalizes the relevance the participant is having in the adapter
+  const _getWeight = (participantHeartbeat: number, totalHeartbeat: number): number => {
```

Comment:
Don't really know what to do with this. It's true that the weight approach is not used, but could be used instead of the even capacity. The difference in the maxAge calculation:
- **Weight**: Each participant will get as much capacity as it needs based on its heartbeat agains the total adapter heartbeat. More heartbeat, more capacity. This has the problem explained in the PR description
- **Even**: Each participant will get the same capacity